### PR TITLE
tree: Allow MapNodes from Records

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Allow `Record` typed object to be used to construct MapNodes.

More details in changeset.

This is one change of several working toward improved JSON compatibility.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: .changeset/green-spies-arrive.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+Before this change, an `Iterable<string, Child>` was required, but now an object like `{key1: Child1, key2: Child2}` is allowed.
+
+Full example using this new API:
+```typescript
+class Schema extends schemaFactory.map("ExampleMap", schemaFactory.number) {}
+const fromRecord = new Schema({ x: 5 });
+```
+
+This new feature makes it possible for schema,
+which do not require unhydrated nodes to differentiate ambiguous unions,
```
Comment (CraigMacomber) [inline .changeset/green-spies-arrive.md:19]: Detection of which schema have this ambiguity issue, and some related details and API which build on the Json support added in this PR are included in https://github.com/microsoft/FluidFramework/pull/22022 as drafts.

File: packages/dds/tree/src/simple-tree/toMapTree.ts [hunk 2]
```diff
@@ -595,10 +603,15 @@ function shallowCompatibilityTest(
 		return mapOrArray ? CompatibilityLevel.Normal : CompatibilityLevel.None;
 	}
 
-	if (mapOrArray) {
+	if (schema.kind === NodeKind.Array) {
 		return CompatibilityLevel.None;
 	}
 
+	if (schema.kind === NodeKind.Map) {
+		// When not unioned with an ObjectNode, allow objects to be used to crete maps.
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/toMapTree.ts:611]:
```suggestion
		// When not unioned with an ObjectNode, allow objects to be used to create maps.
```

File: packages/dds/tree/src/simple-tree/toMapTree.ts [hunk 3]
```diff
@@ -595,10 +603,15 @@ function shallowCompatibilityTest(
 		return mapOrArray ? CompatibilityLevel.Normal : CompatibilityLevel.None;
 	}
 
-	if (mapOrArray) {
+	if (schema.kind === NodeKind.Array) {
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/toMapTree.ts:606]: While you're fixing the typo in this file, could you also move the comment below - the one that says "Assume record-like object" - up here above this line (or wherever appropriate)? I think that will help readers know what is going on here with the array and map schema check.

**Review by noencke — APPROVED** [review]

File: packages/dds/tree/src/simple-tree/toMapTree.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/toMapTree.ts:608]: Fixed. Good catch: its way clearer now.

File: .changeset/green-spies-arrive.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+Before this change, an `Iterable<string, Child>` was required, but now an object like `{key1: Child1, key2: Child2}` is allowed.
+
+Full example using this new API:
+```typescript
+class Schema extends schemaFactory.map("ExampleMap", schemaFactory.number) {}
+const fromRecord = new Schema({ x: 5 });
+```
+
+This new feature makes it possible for schema,
```
Comment (alexvy86) [inline .changeset/green-spies-arrive.md:18]: Rewording suggestion. I think I got the meaning right.
```suggestion
This new feature makes it possible for schemas to construct a tree entirely from JSON compatible objects using their constructors, as long as they do not require unhydrated nodes to differentiate ambiguous unions.
```

**Review by alexvy86 — APPROVED** [review]
Changeset looks good, just left a rewording suggestion.

File: .changeset/green-spies-arrive.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+which do not require unhydrated nodes to differentiate ambiguous unions,
+to construct trees entirely from JSON compatible objects using their constructors.
+
+Due to limitations of TypeScript and recursive types,
+recursive maps to not currently advertize support for this feature in their typing,
+but it does work at runtime.
```
Comment (tylerbutler) [inline .changeset/green-spies-arrive.md:24]:
```suggestion
Due to limitations of TypeScript and recursive types,
recursive maps do not advertise support for this feature in their typing,
but it works at runtime.
```

Should maps be capitalized in this usage?

File: .changeset/green-spies-arrive.md (hunk 4 above)
Comment (CraigMacomber) [reply .changeset/green-spies-arrive.md:18]: If you want to suggest changing multiple lines, you should select multiple lines when making the comment. Applying that suggestion would just replace line 18 with that, which I'm pretty sure is not what you intended.

File: .changeset/green-spies-arrive.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+Full example using this new API:
+```typescript
+class Schema extends schemaFactory.map("ExampleMap", schemaFactory.number) {}
+const fromRecord = new Schema({ x: 5 });
+```
+
+This new feature makes it possible for schema,
+which do not require unhydrated nodes to differentiate ambiguous unions,
+to construct trees entirely from JSON compatible objects using their constructors.
```
Comment (CraigMacomber) [inline .changeset/green-spies-arrive.md:20]:
```suggestion
This new feature makes it possible for schemas to construct a tree entirely from JSON compatible objects using their constructors,
as long as they do not require unhydrated nodes to differentiate ambiguous unions,
or IFluidHandles (which themselevs are not JSON compatible).
```

**Review by tylerbutler — APPROVED** [review]

File: .changeset/green-spies-arrive.md (hunk 4 above)
Comment (CraigMacomber) [reply .changeset/green-spies-arrive.md]: applied slightly reworked version of change separate from this thread. Good rephrasing!

**Pull request merged**